lib/Sema/CSDiagnostics.cpp

```diff
@@ -4262,6 +4262,33 @@ bool ThrowingFunctionConversionFailure::diagnoseAsError() {
   return true;
 }
 
+bool UnecessaryCoecionFailure::diagnoseAsError() {
+  auto expr = dyn_cast<CoerceExpr>(getAnchor());
+  
+  auto diag = [&]() {
```

Comment:
I'm not sure, but that was the first attempt, but I think you can't assign to an `InFlightDiagnostic` like that because this operator is deleted `InFlightDiagnostic &operator=(InFlightDiagnostic &&) = delete;` :(

Reply:
Oops, you're right. I suppose the diagnostics could be put inside the lambda and the checks outside it.

```c++
auto emitCastDiagnostic = [&](bool isTypealias) { ... }
...
bool isTypealias = fromTypeAlias->getDecl() != toTypeAlias->getDecl();
auto diags = emitCastDiagnostic(isTypealias)
diags.fixItRemove(...)
```

Reply:
I would suggest not to overthink it and either attach `fixIt*` calls directly to diagnostics or make it `Optional<InFlightDiagnostic>` and use `if/else`. I like the first option better.